change-validation/src/main/scala/com/normation/plugins/changevalidation/ValidationNeeded.scala

```diff
@@ -0,0 +1,156 @@
+    val start = System.currentTimeMillis()
+    val res = for {
+      groups    <- groupLib.getFullGroupLibrary()
+      nodeInfo  <- nodeInfoService.getAll()
+      monitored <- monitoredTargets()
+    } yield {
+      val targets = Set(change.newRule) ++ change.previousRule.toSet
+      checkNodeTargetByRule(groups, nodeInfo, monitored, targets)
+    }
+    ChangeValidationLogger.Metrics.debug(s"Check rule '${change.newRule.name}' [${change.newRule.id.value}] change requestion need for validation in ${System.currentTimeMillis() - start}ms")
+    res
+  }
+
+  /**
+   * This method checks if at least one of the nodes belonging to rule targets for the change
+   * is supervised.
+   */
+  def checkNodeTargetByRule(groups: FullNodeGroupCategory, allNodeInfo: Map[NodeId, NodeInfo], monitored: Set[SimpleTarget], rules: Set[Rule]): Boolean = {
+    val monitoredNodes = groups.getNodeIds(monitored.map(identity), allNodeInfo)
+    val changes = rules.flatMap(_.targets)
+    val exists = groups.getNodeIds(changes, allNodeInfo).exists(nodeId => monitoredNodes.contains(nodeId))
+    // we want to let the log knows why the change request need validation
+    if(exists && ChangeValidationLogger.isDebugEnabled) {
+      rules.foreach { rule =>
+        groups.getNodeIds(rule.targets, allNodeInfo).find(nodeId => monitoredNodes.contains(nodeId)).foreach { node =>
+           ChangeValidationLogger.debug(s"Node '${node.value}' belongs to both a supervised group and is a target of rule '${rule.name}' [${rule.id.value}]")
+        }
+      }
+    }
+    exists
+  }
+
+  /*
+   * We want to check if the supervised node are the same before and after the change to avoid the case:
+   * - the group (not supervised) does not contain any sypervised nodes,
+   * - the group is applied to a directive,
+   * - the group (still not supervised) is modified to include supervised nodes (because they also
+   *   belong to an other group)
+   * - now the rule is applied to supervised node, but no validation was done.
+   */
+  override def forNodeGroup(actor: EventActor, change: NodeGroupChangeRequest): Box[Boolean] = {
+    // Here we need to test the future content of the group, and not the current one.
+    // So we need to know:
+    // - the list of supervised node in the group before the change,
+    // => non empty means validation needed,
+    // - the list of supervised node in the group after change,
+    // => non empty means validation needed
+
+    val start = System.currentTimeMillis()
+
+    val res = for {
+      groups      <- groupLib.getFullGroupLibrary()
+      allNodeInfo <- nodeInfoService.getAll()
+      monitored   <- monitoredTargets()
+    } yield {
+      val targetNodes = change.newGroup.serverList ++ change.previousGroup.map(_.serverList).getOrElse(Set())
+      val exists = groups.getNodeIds(monitored.map(identity), allNodeInfo).find(nodeId => targetNodes.contains(nodeId))
```

Comment:
Don't we want to use intersect and list all Nodes making a conflict ? could be a problem if there is a lot of Nodes though ...

Reply:
I prefer to just give an example, else we are going to list the union of all nodes in faulty group, so perhaps thousand of them. 
And I use the idea that either the user missed a special case (and there will be few nodes on it, so just an example should be ok), or he forgot a whole group (and in that case, just an example of the node should be ok). 

I could also limit to for ex 10 nodes, so that in the case of special cases, it's likely that they are all displayed.

Reply:
Ok !! I merge this !! let's try that because and we'll see if it's ok or not when using it